Advanced test environment

Artemka374 opened the pull request:
This is PR for this issue: https://github.com/Supercolony-net/openbrush-contracts/pull/136#issuecomment-1167236655
that was started in https://github.com/Supercolony-net/ink/pull/4
- [x] Add support of `call_flags`
- [x] Error & revert handling

Artemka374 committed:
initial change for advanced ink test environment

Artemka374 committed:
initial change for advanced ink test environment

Artemka374 committed:
fix never type

Artemka374 committed:
Merge remote-tracking branch 'origin/feature/advanced-test-environment' into feature/advanced-test-environment

# Conflicts:
#	crates/env/src/api.rs
#	crates/env/src/backend.rs
#	crates/ink/codegen/src/generator/dispatch.rs

Artemka374 committed:
fix bugs

Artemka374 committed:
fix dispatch

Artemka374 commented:
###

Artemka374 committed:
initial implementation of call flags with tests

Artemka374 committed:
fix call_flags bug

coreggon11 commented on crates/engine/src/ext.rs:
```diff
@@ -110,20 +115,75 @@ impl ReturnCode {
+        }
+    }
+
+    pub fn increase_entrance(&mut self, callee: Vec<u8>) -> Result {
```
```suggestion
    pub fn increase_entrance_count(&mut self, callee: Vec<u8>) -> Result {
```

coreggon11 commented on crates/engine/src/ext.rs:
```diff
@@ -110,20 +115,75 @@ impl ReturnCode {
+    }
+
+    pub fn increase_entrance(&mut self, callee: Vec<u8>) -> Result {
+        let entrance = self.entrance_count.get(&callee).unwrap_or(&0).clone();
```
```suggestion
        let entrance_count = self.entrance_count.get(&callee).unwrap_or(&0).clone();
```

coreggon11 replied:
you can use `map_or` (return 0 if None, count + 1 if Some)

coreggon11 commented on crates/engine/src/ext.rs:
```diff
@@ -110,20 +115,75 @@ impl ReturnCode {
+        Ok(())
+    }
+
+    pub fn decrease_entrance(&mut self, callee: Vec<u8>) -> Result {
```
```suggestion
    pub fn decrease_entrance_count(&mut self, callee: Vec<u8>) -> Result {
```

coreggon11 commented on crates/engine/src/ext.rs:
```diff
@@ -110,20 +115,75 @@ impl ReturnCode {
+    }
+
+    pub fn decrease_entrance(&mut self, callee: Vec<u8>) -> Result {
+        let entrance = self.entrance_count.get(&callee).unwrap_or(&0).clone();
```
```suggestion
        let entrance_count = self.entrance_count.get(&callee).unwrap_or(&0).clone();
```

coreggon11 replied:
you can `map_or_else` where you return Err and use `?` operator to return error directly if None, and modify by -1 if Some. delete the entry if it is 0

coreggon11 commented on crates/engine/src/ext.rs:
```diff
@@ -110,20 +115,75 @@ impl ReturnCode {
 }
 
+#[derive(Default)]
+pub struct ContractStore {
```
maybe ContractStorage would be better name?

coreggon11 commented on crates/engine/src/lib.rs:
```diff
@@ -22,7 +22,7 @@ pub mod test_api;
 mod chain_extension;
 mod database;
-mod exec_context;
+pub mod exec_context;
```
I think we dont want to expose this

Artemka374 replied:
but it won't work without it

coreggon11 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -446,12 +477,134 @@ impl TypedEnvBackend for EnvInstance {
+        Ok(return_value)
     }
 
     fn invoke_contract_delegate<E, Args, R>(
```
maybe we can leave this unimplemented

Artemka374 replied:
agree

coreggon11 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -513,31 +710,45 @@ impl TypedEnvBackend for EnvInstance {
         E: Environment,
     {
-        unimplemented!("off-chain environment does not support cross-contract calls")
+        unimplemented!("off-chain environment does not support `seal_caller_is_origin`")
```
what about this?

coreggon11 commented on crates/ink/codegen/src/generator/dispatch.rs:
```diff
@@ -79,12 +79,7 @@ impl GenerateCode for Dispatch<'_> {
             #constructor_decoder_type
             #message_decoder_type
-
-            #[cfg(not(test))]
```
why remove this?

Artemka374 replied:
because I've added that part directly in entry points

Artemka374 committed:
implement revert storage

Artemka374 committed:
minor changes and apply suggestions

Artemka374 committed:
fix bug

Artemka374 committed:
fix bug

coreggon11 commented on crates/engine/src/ext.rs:
```diff
@@ -110,20 +117,81 @@ impl ReturnCode {
+    }
+
+    pub fn set_allow_reentry(&mut self, callee: Vec<u8>, allow: bool) {
+        if allow == false {
```
rewrite this to `if allow { ... } else { ... }`

coreggon11 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -245,14 +247,30 @@ impl EnvBackend for EnvInstance {
+        R: scale::Encode,
+    {
+        if flags.is_reverted() {
+            panic!("the off-chain env does not implement revert in `seal_return_value`")
```
what does this mean?

Artemka374 replied:
in this case we have 2 different return_value, one is for on-chain environment, that returns (), and another is for on-chain, that actually returns !, we just need to implement both

coreggon11 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -245,14 +247,30 @@ impl EnvBackend for EnvInstance {
     {
-        unimplemented!("the off-chain env does not implement `return_value`")
+        if flags.is_reverted() {
+            self.engine.exec_context.borrow_mut().return_flags = 1;
```
maybe we do not need to call the field `return_flags` if we only want to check for revert, let's make it `reverted: bool`

Artemka374 replied:
agree

coreggon11 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -446,12 +477,91 @@ impl TypedEnvBackend for EnvInstance {
-        let _input = params.exec_input();
-        unimplemented!("off-chain environment does not support contract invocation")
+
+        let call_flags = params.call_flags().into_u32();
+
+        let transferred_value = params.transferred_value();
+
+        let caller = self.engine.exec_context.borrow().callee.clone();
```
```suggestion
        let call_flags = params.call_flags().into_u32();
        let transferred_value = params.transferred_value();
        let caller = self.engine.exec_context.borrow().callee.clone();
```

coreggon11 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -446,12 +477,91 @@ impl TypedEnvBackend for EnvInstance {
+        // if the call was reverted, previous one should be reverted too
+        previous_context.return_flags |= self.engine.exec_context.borrow().return_flags;
+
+        let _ = self.engine.exec_context.replace(previous_context);
```
do we need let here?

coreggon11 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -478,12 +588,56 @@ impl TypedEnvBackend for EnvInstance {
+
+        deploy_fn();
+
+        let _ = self.engine.exec_context.replace(previous_context);
```
do we need let here?

coreggon11 commented on crates/env/src/error.rs:
```diff
@@ -49,6 +49,8 @@ pub enum Error {
     LoggingDisabled,
     /// ECDSA pubkey recovery failed. Most probably wrong recovery id or signature.
     EcdsaRecoveryFailed,
+    /// documentation
```
do we need this?

coreggon11 commented on examples/delegator/adder/lib.rs:
```diff
@@ -30,3 +30,40 @@ mod adder {
+    use ink::primitives::Hash;
+
+    #[test]
+    fn it_works() {
```
find a better name for this test

coreggon11 commented on examples/reentrancy/contract2/lib.rs:
```diff
@@ -0,0 +1,66 @@
+#![cfg_attr(not(feature = "std"), no_std)]
```
canw e find a better name for this contract?

Artemka374 committed:
implement caller_is_origin

Artemka374 committed:
apply suggestions

Artemka374 committed:
add README for the reentrancy example and apply suggestions

Artemka374 committed:
Merge branch 'master' into feature/advanced-test-environment

Artemka374 committed:
cargo clippy

Artemka374 committed:
fix clippy

Artemka374 committed:
fix unreachable code

Artemka374 committed:
add allow nonminimal bool

Artemka374 committed:
apply clippy on examples

Artemka374 committed:
apply clippy on delegator example

Artemka374 committed:
fix no_implicit_prelude error

Artemka374 committed:
move tests to fallback_contract

Artemka374 committed:
move fallback_contract to reentrancy dir

Artemka374 committed:
add tests for contract storage

Artemka374 committed:
fix clippy for test

Artemka374 committed:
tail call is not supported yet

ivan770 commented on crates/engine/src/ext.rs:
```diff
@@ -513,6 +605,92 @@ impl Engine {
+        _call_flags: u32,
+        _output: Vec<u8>,
+    ) -> core::result::Result<(), Error> {
+        // let _tail_call = ((call_flags & 4) >> 2) != 0;
```
Are those comments required?

ivan770 commented on crates/env/src/backend.rs:
```diff
@@ -235,16 +240,7 @@ pub trait EnvBackend {
     where
         T: scale::Decode;
 
-    /// Returns the value back to the caller of the executed contract.
```
Documentation regression?

Artemka374 committed:
add generate address method

Artemka374 committed:
implement invoke_delegate

Artemka374 committed:
fix compilation errors

Artemka374 committed:
apply suggestions

Artemka374 committed:
minor code style changes

Artemka374 committed:
wrap engine into Rc<RefCell>

Artemka374 committed:
use `let engine = self.engine.borrow()` where it is possible

Artemka374 committed:
Merge branch 'master' into feature/advanced-test-environment

Artemka374 committed:
Merge branch 'master' into feature/advanced-test-environment

Artemka374 committed:
fix compilation with ink! 4.0.0-beta.1

Artemka374 committed:
fix tests and update test in delegator example

Artemka374 committed:
add unreachable to test to work with our api

Artemka374 committed:
Merge branch 'master' into feature/advanced-test-environment

Artemka374 committed:
overwrite ui tests

Artemka374 commented on crates/engine/src/ext.rs:
```diff
@@ -110,7 +115,66 @@ impl ReturnCode {
+pub struct ContractStorage {
+    pub instantiated: HashMap<Vec<u8>, Vec<u8>>,
+    pub entrance_count: HashMap<Vec<u8>, u32>,
+    pub allow_reentry: HashMap<Vec<u8>, bool>,
```
maybe better to rename it?

Artemka374 commented on crates/engine/src/ext.rs:
```diff
@@ -123,10 +187,13 @@ pub struct Engine {
     pub chain_spec: ChainSpec,
     /// Handler for registered chain extensions.
-    pub chain_extension_handler: ChainExtensionHandler,
+    pub chain_extension_handler: Rc<RefCell<ChainExtensionHandler>>,
```
need to think about using another approach to be able cloning the struct

Artemka374 commented on crates/engine/src/ext.rs:
```diff
@@ -513,6 +582,79 @@ impl Engine {
+    }
+
+    /// Apply call flags for the call and return the input that might be changed
+    pub fn apply_code_flags_before_call(
```
maybe makes sense to rename it since we change entrance_count anyway

Artemka374 commented on crates/engine/src/ext.rs:
```diff
@@ -513,6 +582,79 @@ impl Engine {
+    }
+
+    /// Apply call flags after the call
+    pub fn apply_code_flags_after_call(
```
the same thing

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -180,9 +182,53 @@ impl EnvInstance {
         scale::Decode::decode(&mut &full_scope[..]).map_err(Into::into)
     }
+
+    fn generate_address(
```
maybe we should not leave this in the scope of impl envInstanse because it is not really uses EnvInstance

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -180,9 +182,53 @@ impl EnvInstance {
+    }
+
+    /// Generates new execution context, replaces it as current and returns previous one
+    fn generate_callee_context(
```
The same thing and need to think about renaming

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -233,32 +284,37 @@ impl EnvBackend for EnvInstance {
     }
 
-    fn return_value<R>(&mut self, _flags: ReturnFlags, _return_value: &R) -> !
+    #[cfg(not(all(not(feature = "std"), target_arch = "wasm32")))]
```
why this cfg

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -366,8 +425,21 @@ impl EnvBackend for EnvInstance {
+            .exec_context
+            .callee
+            .clone()
+            .expect("callee is none");
```
why we are using expect here if we can handle error

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -448,12 +522,82 @@ impl TypedEnvBackend for EnvInstance {
+        let call_flags = params.call_flags().into_u32();
+        let transferred_value = params.transferred_value();
+        let caller = self.engine.borrow().exec_context.callee.clone();
+
```
maybe we can make body of this method smaller?

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -465,10 +609,77 @@ impl TypedEnvBackend for EnvInstance {
-        unimplemented!(
-            "off-chain environment does not support delegated contract invocation"
-        )
+        let code_hash = params.code_hash().as_ref().to_vec();
```
size of body?

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -486,20 +697,71 @@ impl TypedEnvBackend for EnvInstance {
         R: ConstructorReturnType<ContractRef>,
     {
-        let _code_hash = params.code_hash();
+        let code_hash = params.code_hash().as_ref().to_vec();
```
size of body?

Artemka374 commented on crates/env/src/engine/off_chain/impls.rs:
```diff
@@ -509,43 +771,72 @@ impl TypedEnvBackend for EnvInstance {
+            .exec_context
+            .origin
+            .clone()
+            .expect("stack should not be empty")
```
is it the best naming for error?

Artemka374 commented on examples/reentrancy/lib.rs:
```diff
@@ -0,0 +1,115 @@
+        }
+
+        #[ink(message)]
+        pub fn get_key(&self) -> Key {
```
why do we need it

Artemka374 commented on examples/reentrancy/lib.rs:
```diff
@@ -0,0 +1,115 @@
+    #[ink(storage)]
+    pub struct FallbackContract {
+        /// Stores a single `bool` value on the storage.
+        value: u32,
```
why do we need value here

Artemka374 commented on examples/reentrancy/lib.rs:
```diff
@@ -0,0 +1,115 @@
+
+        #[ink(message, selector = _)]
+        pub fn fallback(&mut self) {
+            ink::env::set_contract_storage(
```
is it necessary?

Artemka374 commented on examples/reentrancy/main_contract/lib.rs:
```diff
@@ -0,0 +1,107 @@
+        }
+
+        #[ink(message)]
+        pub fn get_key(&self) -> Key {
```
remove it

Artemka374 committed:
delete unused methods in examples

Artemka374 committed:
update ui tests

Artemka374 committed:
update readme

123oli commented:
While calling `fallable` from Fallback Contract I got `Contract trapped during execution`. Why MainContract::inc: NotCallable' isn't callable.
<img width="1440" alt="Screenshot 2023-01-28 at 10 56 46" src="https://user-images.githubusercontent.com/42571338/215243434-7e36b5c8-9f97-43cb-bae6-f5c08a2de919.png">

coreggon11 commented:
@123oli did you deploy the main contract? Is the account you are calling indeed the main contract's account id?

123oli commented:
> @123oli did you deploy the main contract? Is the account you are calling indeed the main contract's account id?

Yes, I have deployed main contract as well and while deploying fallable contract I indeed choose main contract's account id.

Artemka374 commented:
@123oli but actually you can't just set the callee as AccountId for fallback contract, because as you can see, inside of contract it's type is `MainContractRef`, anyway that's a good point to change the way it works, so everybody could test it on-chain

Artemka374 committed:
Change the way contracts are instantiated in example, update readme

Artemka374 committed:
fix clippy

Artemka374 commented:
@123oli but it is still AccountId, you can see the type of variable in UI. Also, I've changed the way contracts are deployed and added README, so you can use guide from there. But if you try to just call fallback as you described previously, it won't work, because reentrancy will occur in fallback contract, which is not allowed, so you can try to just call main contract first, and then you will be able to call fallback method directly.

Artemka374 committed:
Merge branch 'master' into feature/advanced-test-environment

Artemka374 committed:
update tests to new ink! version
